src/cdomains/intDomain.ml

```diff
@@ -1861,193 +1861,247 @@ module Booleans = MakeBooleans (
+    in
+    match a, b with
+    | Inc x, Inc y -> ISet.compare x y
+    | Exc (x,r), Exc (y, s) ->
+        let comp_sets = ISet.compare x y in
+        if comp_sets = 0 then R.compare r s else comp_sets
+    | _, _ -> compare (value a) (value b)
+
+  let hash = function
+    | Inc x ->
+      let m = match BI.to_int (ISet.min_elt x) with n -> n | exception Not_found -> 41 | exception Failure _ -> 61 in (* Failure occurs when the conversion to int did not work, Not_found occurs when the set is empty *)
+      let n = match BI.to_int (ISet.max_elt x) with n -> n | exception Not_found -> 43 | exception Failure _ -> 67 in
+      3  * ISet.cardinal x + 7 * m + 13 * n
+    | Exc (x, r) ->
+      let m = match BI.to_int (ISet.min_elt x) with n -> n | exception Not_found  -> 53 | exception Failure _ -> 71 in
+      let n = match BI.to_int (ISet.max_elt x) with n -> n | exception Not_found  -> 59 | exception Failure _ -> 73 in
+      R.hash r + 5  * ISet.cardinal x + 11 * m + 19 * n
+
+  let norm ikind v =
+    let min, max = min_int ikind, max_int ikind in
+    (* Whether the value v lies within the values of the specified ikind. *)
+    let value_in_ikind v =
+      I.compare min v <= 0 && I.compare v max <= 0
+    in
+    (* Whether the range r lies witihin the range of the ikind. *)
+    let range_in_ikind r =
+      R.leq r (size ikind)
+    in
+    match v with
+    | Inc xs when ISet.for_all value_in_ikind xs -> v
+    | Exc (xs, r) when ISet.for_all value_in_ikind xs && range_in_ikind r -> v
+    | _ -> top_of ikind
+
   let equal_to i = function
     | Inc x ->
-      if List.mem i x then
-        if List.length x = 1 then `Eq
+      if ISet.mem i x then
+        if ISet.is_singleton x then `Eq
         else `Top
       else `Neq
     | Exc (x, r) ->
-      if List.mem i x then `Neq
+      if ISet.mem i x then `Neq
       else `Top
-  let of_int x = Inc [x]
 
-  let top_if_not_in_int64 ik f x = try f x with Size.Not_in_int64 -> top_of ik
-  let cast_to ?torg ik = top_if_not_in_int64 ik @@ function
+  let cast_to ?torg ik v = norm ik @@ match v with
     | Exc (s,r) ->
       let r' = size ik in
       if R.leq r r' then (* upcast -> no change *)
         Exc (s, r)
       else if torg = None then (* same static type -> no overflows for r, but we need to cast s since it may be out of range after lift2_inj *)
-        let s' = List.map (I.cast_to ik) s in
+        let s' = ISet.map (I.cast_to ik) s in
         Exc (s', r')
       else (* downcast: may overflow *)
-        Exc ([], r')
-    |  Inc x -> Inc (List.map (Integers.cast_to ik) x)
+        Exc (ISet.empty, r')
+    |  Inc x -> Inc (ISet.map (BigInt.cast_to ik) x)
+
+  let of_int ikind x = cast_to ikind (Inc (ISet.singleton x))
 
-  let of_interval ik (x,y) = (* TODO this implementation might lead to very big lists; also use ana.int.enums_max? *)
+  let of_interval ik (x,y) = (* TODO this implementation might lead to very big lists; use some configurable limit? *)
```

Comment:
When does `of_interval` actually get used? It could be a problem if somehow we take a large interval based on the interval analysis and then try to re-express it into enums. In such case it might only make sense to handle singleton intervals exactly and everything else as top.

Reply:
The `of_interval` function in Enums is actually never called. 
Only the `of_interval` function of the interval domain gets actually called, in particular to construct ranges in `DefExc` and `Enums`.

Reply:
I changed the implementation now anyway to only produce singleton sets or top.